typescript/sdk/src/ism/EvmIsmModule.ts

```diff
@@ -1,89 +1,727 @@
+
+    this.chainName = this.multiProvider.getChainName(this.args.chain);
+    this.domainId = this.multiProvider.getDomainId(this.chainName);
   }
 
   public async read(): Promise<IsmConfig> {
-    return await this.reader.deriveIsmConfig(this.args.addresses.deployedIsm);
+    return this.reader.deriveIsmConfig(this.args.addresses.deployedIsm);
   }
 
-  public async update(config: IsmConfig): Promise<EthersV5Transaction[]> {
-    throw new Error('Method not implemented.');
+  // whoever calls update() needs to ensure that targetConfig has a valid owner
+  public async update(
+    targetConfig: IsmConfig,
+  ): Promise<AnnotatedEthersV5Transaction[]> {
+    const currentConfig =
+      typeof this.args.config === 'string'
+        ? this.args.addresses.deployedIsm
+        : await this.read();
 
-    const destination = this.multiProvider.getChainName(this.args.chain);
-    await this.creator.update({
-      destination,
-      config,
-      existingIsmAddress: this.args.addresses.deployedIsm,
+    // Update the config in case it's a custom ISM
+    this.args.config = targetConfig;
+
+    // moduleMatchesConfig expects any domain filtering to have been done already
+    if (
+      typeof targetConfig !== 'string' &&
+      (targetConfig.type === IsmType.ROUTING ||
+        targetConfig.type === IsmType.FALLBACK_ROUTING)
+    ) {
+      // filter for known domains
+      const { availableDomains } = this.filterRoutingIsmDomains({
+        config: targetConfig,
+      });
+      targetConfig.domains = availableDomains;
+    }
+
+    // this calls resolveOrDeployAccountOwner under the hood
+    const configMatches = await moduleMatchesConfig(
+      this.chainName,
+      this.args.addresses.deployedIsm,
+      targetConfig,
+      this.multiProvider,
+      this.factories,
+      this.args.addresses.mailbox,
+    );
+
+    // If configs match, no updates needed
+    if (configMatches) {
+      return [];
+    }
+
+    // Else, we have to figure out what an update for this ISM entails
+
+    // If target config is a custom ISM, just update the address
+    // if config -> custom ISM, update address
+    // if custom ISM -> custom ISM, update address
+    if (typeof targetConfig === 'string') {
+      // TODO: https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3773
+      this.args.addresses.deployedIsm = targetConfig;
+      return [];
+    }
+
+    // Check if we need to deploy a new ISM
+    if (
+      // if custom ISM -> config, do a new deploy
+      typeof currentConfig === 'string' ||
+      // if config -> config, AND types are different, do a new deploy
+      currentConfig.type !== targetConfig.type ||
+      // unless the new ISM is a Routing or Pausable ISM, do a new deploy
+      (targetConfig.type !== IsmType.ROUTING &&
+        targetConfig.type !== IsmType.FALLBACK_ROUTING &&
+        targetConfig.type !== IsmType.PAUSABLE)
+    ) {
+      const contract = await this.deploy({
+        config: targetConfig,
+      });
```

Comment:
I would have expected `this.deploy` to handle reuse of addresses on the module

Reply:
not sure what you mean by this

Reply:
make this into a "mutable ism" subset